storage/integrate.go

```diff
@@ -0,0 +1,337 @@
+	newRoot, err := baseRange.GetRootHash(nil)
+	if err != nil {
+		return 0, nil, nil, fmt.Errorf("failed to calculate new root hash: %w", err)
+	}
+
+	if err := tc.Err(); err != nil {
+		return 0, nil, nil, err
+	}
+
+	// All calculation is now complete, all that remains is to store the new
+	// tiles and updated log state.
+	klog.V(1).Infof("New log state: size 0x%x hash: %x", baseRange.End(), newRoot)
+
+	return baseRange.End(), newRoot, tc.Tiles(), nil
+
+}
+
+// tileReadCache is a structure which provides a very simple thread-safe map of tiles.
+type tileReadCache struct {
+	sync.RWMutex
+
+	entries map[string]*populatedTile
+}
+
+func newTileReadCache() tileReadCache {
+	return tileReadCache{
+		entries: make(map[string]*populatedTile),
+	}
+}
+
+// Get returns a previously set tile and true, or, if no such tile is in the cache, returns nil and false.
+func (r *tileReadCache) Get(tileID TileID, treeSize uint64) (*populatedTile, bool) {
+	r.RLock()
+	defer r.RUnlock()
+	k := layout.TilePath(uint64(tileID.Level), tileID.Index, treeSize)
+	e, ok := r.entries[k]
+	return e, ok
+}
+
+// Set associates the given tileID with a tile.
+func (r *tileReadCache) Set(tileID TileID, treeSize uint64, t *populatedTile) {
+	r.Lock()
+	defer r.Unlock()
+	k := layout.TilePath(uint64(tileID.Level), tileID.Index, treeSize)
+	if e, ok := r.entries[k]; ok && !e.Equals(t) {
+		klog.Infof("OVERWRITE TILE %v:\nExisting:%x\nNew\n%x", k, e, t)
+		panic(fmt.Errorf("Attempting to overwrite %v with different content", k))
+	}
+	r.entries[k] = t
+}
+
+// tileWriteCache is a simple cache for storing the newly created tiles produced by
+// the integration of new leaves into the tree.
+//
+// Calls to Visit will cause the map of tiles to become filled with the set of
+// `dirty` tiles which need to be flushed back to storage to preserve the updated
+// tree state.
+//
+// Note that by itself, this cache does not update any persisted state.
+type tileWriteCache struct {
+	sync.Mutex
+	m   map[TileID]*populatedTile
+	err []error
+
+	treeSize uint64
+	getTile  getPopulatedTileFunc
+}
+
+// newtileWriteCache creates a new cache for the given treeSize, and uses the provided
+// function to fetch existing tiles which are being updated by the Visitor func.
+func newTileWriteCache(treeSize uint64, getTile getPopulatedTileFunc) *tileWriteCache {
+	return &tileWriteCache{
+		m:        make(map[TileID]*populatedTile),
+		treeSize: treeSize,
+		getTile:  getTile,
+	}
+}
+
+// Err returns an aggregated view of any errors seen by the visitor function.
+//
+// This can be used to check whether updates to the tile cache made by the visitor
+// were made correctly. Any errors returned here are most likely to be due to
+// the cache attempting to read an existing tile which is being updated.
+func (tc *tileWriteCache) Err() error {
+	return errors.Join(tc.err...)
+}
+
+// Visit should be called once for each newly set non-ephemeral node in the
```

Comment:
nit: This comment should start with `Visitor ...`. It appears to jump over what this function does and talks about how its return value should be used.

Reply:
Updated